powerauth-push-server/src/main/java/io/getlime/push/controller/rest/PushDeviceController.java

```diff
@@ -93,17 +95,32 @@ public Response createDevice(@RequestBody ObjectRequest<CreateDeviceRequest> req
         }
         logger.info("Received createDevice request, app ID: {}, activation ID: {}, token: {}, platform: {}", requestObject.getAppId(),
                 requestObject.getActivationId(), maskPushToken(requestObject.getToken()), requestObject.getPlatform());
-        String errorMessage = CreateDeviceRequestValidator.validate(requestObject);
+        final String errorMessage = CreateDeviceRequestValidator.validate(requestObject);
         if (errorMessage != null) {
             throw new PushServerException(errorMessage);
         }
-        String appId = requestObject.getAppId();
-        String pushToken = requestObject.getToken();
-        String platform = requestObject.getPlatform();
-        String activationId = requestObject.getActivationId();
-        List<PushDeviceRegistrationEntity> devices = lookupDeviceRegistrations(appId, activationId, pushToken);
-        final AppCredentialsEntity appCredentials = findAppCredentials(appId);
-        PushDeviceRegistrationEntity device;
+
+        final AppCredentialsEntity appCredentials = findAppCredentials(requestObject.getAppId());
+
+        // In case of parallel requests to create new device, the createOrUpdateDevice may fail because of unique
+        // constraint violation. Try to repeat the save.
+        RetryTemplate.builder()
+                .retryOn(DataIntegrityViolationException.class)
+                .build()
+                .execute(ctx -> createOrUpdateDevice(requestObject, appCredentials));
+
+        logger.info("The createDevice request succeeded, app ID: {}, activation ID: {}, platform: {}", requestObject.getAppId(), requestObject.getActivationId(), requestObject.getPlatform());
+        return new Response();
+    }
+
+    private Void createOrUpdateDevice(final CreateDeviceRequest requestObject, final AppCredentialsEntity appCredentials) throws PushServerException {
```

Comment:
Are you sure this method should return `Void` and not `void`?

Reply:
Indeed, `Void` is correct, since `public final <T, E extends Throwable> T execute(RetryCallback<T, E> retryCallback) throws E {`

Reply:
Ah, ok, my bad.